src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -1756,32 +1756,50 @@ public static decimal Sum<TSource>(this IEnumerable<TSource> source, Func<TSourc
         public static int Min(this IEnumerable<int> source)
         {
             if (source == null) throw Error.ArgumentNull("source");
-            int value = 0;
-            bool hasValue = false;
-            foreach (int x in source)
+            int value;
+            using (IEnumerator<int> e = source.GetEnumerator())
             {
-                if (hasValue)
+                if (!e.MoveNext()) throw Error.NoElements();
+                value = e.Current;
+                while (e.MoveNext())
                 {
+                    int x = e.Current;
                     if (x < value) value = x;
                 }
-                else
-                {
-                    value = x;
-                    hasValue = true;
-                }
             }
-            if (hasValue) return value;
-            throw Error.NoElements();
+            return value;
         }
 
         public static int? Min(this IEnumerable<int?> source)
         {
             if (source == null) throw Error.ArgumentNull("source");
-            int? value = null;
-            foreach (int? x in source)
+            int? value;
+            using (IEnumerator<int?> e = source.GetEnumerator())
             {
-                if (value == null || x < value)
-                    value = x;
+                if (!e.MoveNext()) return null;
+                value = e.Current;
```

Comment:
Something like:
```C#
                do
                {
                    if (!e.MoveNext()) goto exit; // goes to "return value"
                    value = e.Current;
                }
                while (!value.HasValue);
```
I did not time it. It should result in a more compact code, but only measuring can tell if that leads to faster execution.